File: ios/FoodMatch/Models/Restaurant.swift
import Foundation
import CoreLocation

// MARK: - Restaurant Model
struct Restaurant: Identifiable, Codable, Equatable {
    let id: String
    let name: String
    let imageURL: String?
    let rating: Double
    let reviewCount: Int
    let price: String?
    let categories: [Category]
    let location: Location
    let phone: String?
    let distance: Double?
    let isClosed: Bool
    let url: String?
    
    // For SwiftUI List identification
    static func == (lhs: Restaurant, rhs: Restaurant) -> Bool {
        return lhs.id == rhs.id
    }
}

// MARK: - Category Model
struct Category: Codable, Identifiable {
    let id = UUID()
    let alias: String
    let title: String
    
    private enum CodingKeys: String, CodingKey {
        case alias, title
    }
}

// MARK: - Location Model  
struct Location: Codable {
    let address1: String?
    let address2: String?
    let address3: String?
    let city: String?
    let zipCode: String?
    let country: String?
    let state: String?
    let displayAddress: [String]
    
    private enum CodingKeys: String, CodingKey {
        case address1, address2, address3, city, country, state
        case zipCode = "zip_code"
        case displayAddress = "display_address"
    }
    
    var formattedAddress: String {
        return displayAddress.joined(separator: ", ")
    }
}

// MARK: - Yelp API Response Models
struct YelpSearchResponse: Codable {
    let businesses: [YelpBusiness]
    let total: Int
    let region: YelpRegion?
}

struct YelpBusiness: Codable {
    let id: String
    let alias: String?
    let name: String
    let imageUrl: String?
    let isClosed: Bool
    let url: String?
    let reviewCount: Int
    let categories: [YelpCategory]
    let rating: Double
    let coordinates: YelpCoordinates
    let transactions: [String]
    let price: String?
    let location: YelpLocation
    let phone: String?
    let displayPhone: String?
    let distance: Double?
    
    private enum CodingKeys: String, CodingKey {
        case id, alias, name, url, categories, rating, coordinates, transactions, price, location, phone
        case imageUrl = "image_url"
        case isClosed = "is_closed"
        case reviewCount = "review_count"
        case displayPhone = "display_phone"
        case distance
    }
    
    // Convert YelpBusiness to Restaurant
    func toRestaurant() -> Restaurant {
        return Restaurant(
            id: id,
            name: name,
            imageURL: imageUrl,
            rating: rating,
            reviewCount: reviewCount,
            price: price,
            categories: categories.map { Category(alias: $0.alias, title: $0.title) },
            location: Location(
                address1: location.address1,
                address2: location.address2,
                address3: location.address3,
                city: location.city,
                zipCode: location.zipCode,
                country: location.country,
                state: location.state,
                displayAddress: location.displayAddress
            ),
            phone: displayPhone,
            distance: distance,
            isClosed: isClosed,
            url: url
        )
    }
}

struct YelpCategory: Codable {
    let alias: String
    let title: String
}

struct YelpCoordinates: Codable {
    let latitude: Double
    let longitude: Double
}

struct YelpLocation: Codable {
    let address1: String?
    let address2: String?
    let address3: String?
    let city: String?
    let zipCode: String?
    let country: String?
    let state: String?
    let displayAddress: [String]
    
    private enum CodingKeys: String, CodingKey {
        case address1, address2, address3, city, country, state
        case zipCode = "zip_code"
        case displayAddress = "display_address"
    }
}

struct YelpRegion: Codable {
    let center: YelpCoordinates
} 

File: web/src/types/Restaurant.ts
export interface Restaurant {
  id: string;
  name: string;
  imageUrl: string;
  rating: number;
  reviewCount: number;
  categories: string[];
  price: string;
  phone: string;
  address: string;
  city: string;
  state: string;
  zipCode: string;
  latitude: number;
  longitude: number;
  distance: number;
  isOpen: boolean;
  url: string;
}

export interface SwipeDirection {
  LIKE: 'like';
  PASS: 'pass';
}

export interface SwipeResult {
  restaurantId: string;
  direction: 'like' | 'pass';
  timestamp: Date;
}

export interface Match {
  id: string;
  restaurant: Restaurant;
  users: string[];
  timestamp: Date;
}

export interface YelpSearchResponse {
  businesses: YelpBusiness[];
  total: number;
  region: {
    center: {
      latitude: number;
      longitude: number;
    };
  };
}

export interface YelpBusiness {
  id: string;
  alias: string;
  name: string;
  image_url: string;
  is_closed: boolean;
  url: string;
  review_count: number;
  categories: Array<{
    alias: string;
    title: string;
  }>;
  rating: number;
  coordinates: {
    latitude: number;
    longitude: number;
  };
  transactions: string[];
  price?: string;
  location: {
    address1: string;
    address2?: string;
    address3?: string;
    city: string;
    zip_code: string;
    country: string;
    state: string;
    display_address: string[];
  };
  phone: string;
  display_phone: string;
  distance: number;
} 